### DISCUSSION TITLE
Fastify plugin function - please help ...

### DISCUSSION THREAD

**Discussion by Rimokas** [fastify/fastify#4408] [Q&A] [answered]
Hi,
It's mine first app with Fastify. 
Can't understand what is mine mistakes with plugin ... 
Here is code. main app.js :

```js
       const path = require( 'path' )
       const fastify = require( 'fastify' )( { logger: false } )
       const lgn = require( './logas' ) // mine login js file
       const conStr = 'mysql://root:pasw@localhost:3306/mineDB'
      
      fastify.register( require( '@fastify/static' ), 
      {
         root: path.join( __dirname, 'public' ),
         prefix: '/public/'
      })
      
      fastify.register( require( '@fastify/formbody' ) )
      fastify.register( require( '@fastify/mysql' ), { connectionString: conStr } )
      fastify.register( lgn )
      
      fastify.get( '/', ( req, res ) => 
      {
         res.sendFile( 'login.html' )
      })
      
      fastify.post( '/login', ( req, res ) => 
      {
         const vart = req.body.user
         const slap = req.body.password
         const ret  = fastify.prisijungt( { vrt: vart, slp: slap } )
         if ( ret )
            res.sendFile( 'index.html' )
      })
      
      fastify.get( '/about', ( req, res ) => 
      {
         res.sendFile( 'about.html' )
      })
      
      fastify.listen({ port: 8080 }, err => 
      {
         if ( err ) throw err
         console.log( `Server run - ${fastify.server.address().port}` )
      })
```
logas.js file : 

```js
function prisijungt( fastify, prm, done )
      {
         const qryStr = "SELECT * FROM users WHERE usr_id = ? and usr_psw = ? and allw = 1 LIMIT 1;"
         fastify.mysql.query( qryStr, [ prm.vrt, prm.slp ], function onResult ( err, result ) 
         {
            return ( err || result )
         })
         done()
      }
      
      module.exports = prisijungt
```

Mine problem is that _ get an error on 'post', where is "fastify.prisijungt" function, registered as plugin : 
{"statusCode":500,"error":"Internal Server Error","message":"fastify.prisijungt is not a function"}

What is wrong here ? Please help ... :(

Many thanks in advance!

**Comment by climba03003** [comment]
Please use `fastify.decorate` to add property to `fastify`.

**Comment by Rimokas** [comment]
Thanks for the answer!

Ok, tried that from samples in internet : 

      `   
      module.exports = (fastify, prm, next ) => 
      {
         fastify.decorate( 'prisijungt', {
         name: 'logas'
         })
         const qryStr = "SELECT * FROM users WHERE usr_id = ? and usr_psw = ? and allw = 1 LIMIT 1;"
         fastify.mysql.query( qryStr, [ prm.vrt, prm.slp ], function onResult ( err, result ) 
         {
            return ( err || result )
         })
         next()
      }`

The same error ... :(

Can you share with some sample ?

**Reply by climba03003** [reply]
Missing a wrapper of `fastify-plugin`.

**Comment by Rimokas** [comment]
I'll be very thankful for any source sample ... :)

**Reply by climba03003** [reply]
Learn the concept first.
https://www.fastify.io/docs/latest/Reference/Encapsulation/

**Reply by climba03003** [reply]
Then, check how to write a proper plugin.
https://www.fastify.io/docs/latest/Guides/Plugins-Guide/

Please scroll to `How to handle encapsulation and distribution` if you only need direct solution.

**Comment by Rimokas** [comment]
Now I found that - https://blog.rafaelgss.com.br/fastify-why-another-framework-js 
No wrapper, 2 plugins. Tried somethig similar - but without success ... 
Something can't understand - so I'm asking for help ...

**Comment by Rimokas** [comment]
Maybe I'm stupid, but after reading that I can't to do that work ... :(

Of course I can query from mysql inside post request - and that is working. But ... the plugins sooner or later it will be required ... 

Can somebody to correct mine piece of source and explain what is wrong here ? I'm in learning process of Fastify - but documentation in most is very elementary - plugins in the same js. If in the outside js - very simplest functions ... 
Mine purpose too very simple - from server let function ( it must be plugin ) from other js to do some stuffs ( here to return results from mysql ) . I tried it in very variants ... 

How to call plugin with mysql query and return results ?

Thanks in advance ...

**Comment by jsumners** [comment]
A plugin is used to decorate the Fastify objects (server, request, reply). Your "plugin" is a one time function that does a query and returns. All of our documentation clearly shows how to write plugins that decorate the Fastify objects with reusable functions.

**Comment by Rimokas** [comment, marked as answer]
Jsummers, thank for the answer ... 

Read this -  https://www.fastify.io/docs/latest/Reference/Plugins/ 

At the end is a sample ... 
Did own logas.js 

      `  const fp = require( 'fastify-plugin' )
      module.exports = fp( function( fastify, prm, done )
      {
         fastify.decorate( 'jungtis', function()
         {
               const qryStr = "SELECT * FROM usersWHERE usr_id = ? and usr_psw = ? and allw = 1 LIMIT 1;"
               fastify.mysql.query( qryStr, [ prm.vrt, prm.slp ], function onResult ( err, result ) 
               {
                  return ( err || result )
               })
            })
         done()
      })
      ` 
Now it's working :) ... Tfuuuu ... I can say that at this moment fully don't understand how and why ... :) . But after some time maybe it will more understandly 

Anyway thanks ...

**Reply by melroy89** [reply]
so this works because you are using the plugin system of Fastify. And decorate is called once, so you can optionally also hook on events like preHandle (but not required in your case). IN your case you use decorate to add a function, again once during start-up. 
After that, you can use `jungtis` to call this method across other places eg. in handle/controller.

Finally, the `fastify-plugin` is required if you want to have good encapsulation across all requests, you can usage this fastify-plugin wrapper. 

More info: https://fastify.dev/docs/latest/Reference/Decorators/#usage
